lib/expressions/Term.ts

```diff
@@ -91,38 +96,57 @@ export class Literal<T> extends Term {
 }
 
 export class NumericLiteral extends Literal<number> {
-  private static readonly specificFormatters: {[key in C.PrimitiveNumericType]: (val: number) => string } = {
+  private readonly specificFormatter: ((val: number) => string);
+  public constructor(
+    public typedValue: number,
+    public typeURL: RDF.NamedNode,
+    public strValue?: string,
+    public language?: string,
+  ) {
+    super(typedValue, typeURL, strValue, language);
+    if (!typeCanBeProvidedTo(<LiteralTypes> typeURL.value, TypeAlias.SPARQL_NUMERIC)) {
+      throw new Error(
+        `TypeUrl '${this.type}' provided to NumericLiteral should implement ${TypeAlias.SPARQL_NUMERIC}`,
+      );
+    }
+
+    const type = this.type;
     // Avoid emitting non lexical integers
-    integer: value => value.toFixed(0),
-    float: value => value.toString(),
-    decimal: value => value.toString(),
-    // // Be consistent with float
-    // decimal: (value) => {
-    //   const jsDecimal = value.toString();
-    //   return jsDecimal.match(/\./)
-    //     ? jsDecimal
-    //     : jsDecimal + '.0';
-    // },
-
-    // https://www.w3.org/TR/xmlschema-2/#double
-    double(value) {
-      const jsExponential = value.toExponential();
-      const [ jsMantisse, jsExponent ] = jsExponential.split('e');
-
-      // Leading + must be removed for integer
-      // https://www.w3.org/TR/xmlschema-2/#integer
-      const exponent = jsExponent.replace(/\+/u, '');
-
-      // SPARQL test suite prefers trailing zero's
-      const mantisse = jsMantisse.includes('.') ?
-        jsMantisse :
-        `${jsMantisse}.0`;
-
-      return `${mantisse}E${exponent}`;
-    },
-  };
-
-  public type: C.PrimitiveNumericType;
+    if (typeCanBeProvidedTo(type, TypeURL.XSD_INTEGER)) {
+      this.specificFormatter = value => value.toFixed(0);
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_DECIMAL)) {
+      this.specificFormatter = value => value.toString();
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_FLOAT)) {
+      this.specificFormatter = value => value.toString();
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_DOUBLE)) {
+      // https://www.w3.org/TR/xmlschema-2/#double
+      this.specificFormatter = value => {
+        const jsExponential = value.toExponential();
+        const [ jsMantisse, jsExponent ] = jsExponential.split('e');
+
+        // Leading + must be removed for integer
+        // https://www.w3.org/TR/xmlschema-2/#integer
+        const exponent = jsExponent.replace(/\+/u, '');
+
+        // SPARQL test suite prefers trailing zero's
+        const mantisse = jsMantisse.includes('.') ?
+          jsMantisse :
+          `${jsMantisse}.0`;
+
+        return `${mantisse}E${exponent}`;
+      };
+      // // Be consistent with float
+      // decimal: (value) => {
+      //   const jsDecimal = value.toString();
+      //   return jsDecimal.match(/\./)
+      //     ? jsDecimal
+      //     : jsDecimal + '.0';
+      // },
```

Comment:
Right, about this, what doe it mean? It's kind of vague for an outsider, could you elaborate on it? Then I will also have a clearer view on what exactly to do with it.

Reply:
The SPARQL spec defines a way to format floats, the JS spec does the same. These aren't necessarily compatible. I can't recall why this is just commented out (it's just adds a trailing `.0` if the number is round), but it might be hint we don't have sufficient tests yet that the way we format things agrees with the SPARQL spec (because we use JS defaults).

Reply:
I agree, that's also a reason I commented https://github.com/comunica/sparqlee/pull/103#discussion_r688370200 .

Reply:
We might need to think about creating js Classses representing the various datatypes? I don't know if this would break things, if it would be a lot of work or if it's even possible. I do think it can be interesting for formatting purposes. In addition, our TypeUrl currently only extend 1 thing. That means we could map this perfectly to js class extension? Wouldn't that be great?

Reply:
@rubensworks left me this: https://github.com/w3c/rdf-tests/issues/58  and that points to #28 . I think this refers to the code here? That would mean we can just remove the comment and there is no reason to handle it?

Reply:
Makes sense, let's clean it up.